### PR TITLE
feat(permissions): add "--deny-*" flags

### LINKED ISSUE #18804: Deny permissions
Right now I run trusted scripts using `deno run -A` but since the list of permissions in the future may change, it may be beneficial to do something like `deno run --allow-all --deny-env` to allow the script to do everything we normally allow it to do but not let it access the environment.

### PR DESCRIPTION
**Pull request by iuioiua**
Closes #18804.

### CODE REVIEW

**Comment by iuioiua** [comment]
Actually, @nayeemrmn, would you be able to implement your new proposed API for `UnaryPermission<EnvDescriptor>`? You have a deeper understanding of the permission system, and it'd be best if you were the one to figure out the technicalities and nuances. Once we have `--deny-env` working, I'll likely be able to follow your pattern and make the same changes to the other permissions. WDYT?

**Comment by nayeemrmn** [comment]
https://github.com/nayeemrmn/deno/commit/97b7048b8828e83a9f4ba8217fce936d7bc089bb Implemented struct and method changes for env and read permissions (temporarily in `UnaryPermission2` while the other exists). It moves as much as possible to generic `UnaryPermission2`, so little has to be done in `impl UnaryPermission2<ReadDescriptor>` etc.

**Comment by iuioiua** [comment]
Excellent! Thank you.

**Comment by iuioiua** [comment]
@nayeemrmn, I've done FFI and write permissions for that PR, which mostly works with my local scripts. However, tests around the place are failing. I'm not super-experienced with the permission system and Rust yet. Are you able to do a quick review of its current state? I'll continue working through the rest, meanwhile.

File: cli/args/flags.rs
Comment (bartlomieju) [inline cli/args/flags.rs:2025]:
```suggestion
  "Deny file system read access. Optionally specify denied paths.\n",
```

File: runtime/permissions/mod.rs
Comment (bartlomieju) [inline runtime/permissions/mod.rs:51]: Let's keep explicit numbers here

File: runtime/permissions/mod.rs
Comment (bartlomieju) [inline runtime/permissions/mod.rs:236]: This is a bit strange. Shouldn't it be `self > other`?

File: runtime/permissions/mod.rs
Comment (bartlomieju) [inline runtime/permissions/mod.rs:1488]: This could use `let Some() else { return ... }` guards for better readability

File: runtime/permissions/mod.rs
Comment (bartlomieju) [inline runtime/permissions/mod.rs:1475]: Ditto

File: runtime/permissions/mod.rs
Comment (bartlomieju) [inline runtime/permissions/mod.rs:1561]: Ditto

File: runtime/permissions/mod.rs
Comment (bartlomieju) [inline runtime/permissions/mod.rs:1580]: Ditto

**Review by bartlomieju — COMMENTED** [review]
This is looking good, nice work @iuioiua and @nayeemrmn.

Let's get some tests going (both unit tests in Rust and integration tests for runtime `Deno.permissions` API) for better visibility where we are right now.

Note to self: manual needs to be updated before landing this PR.

**Comment by iuioiua** [comment]
I want to get current permission tests to pass, but they're not. There's likely something wrong with the `NetDescriptor` implementation.
```
// cargo test --package deno_runtime --lib -- permissions::tests --nocapture

failures:
    permissions::tests::check_specifiers
    permissions::tests::test_check_net_url
    permissions::tests::test_check_net_with_values
    permissions::tests::test_request
    permissions::tests::test_revoke
```

**Comment by iuioiua** [comment]
Some assistance getting the existing tests working would be great. I'll continue to work through new tests either way.

**Comment by bartlomieju** [comment]
@iuioiua I'll look into the tests at first convenience

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [inline runtime/permissions/mod.rs:513]:
```suggestion

  fn stronger_than(&self, other: &Self) -> bool {
    self.0 == other.0 && (self.1 == None || self.1 == other.1)
  }
}
```

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [inline runtime/permissions/mod.rs:404]: My mistake in the implementation I gave, `no_prompt` should be an downwards recursive check
```suggestion
  fn is_no_prompt(&self, desc: &Option<T>) -> bool {
    match desc.as_ref() {
      Some(desc) => self.no_prompt_list.iter().any(|v| desc.stronger_than(v)),
      None => self.no_prompt_global || !self.no_prompt_list.is_empty(),
    }
  }
```

**Comment by iuioiua** [comment]
I've renamed `no_prompt_(global/list)` to `no_to_prompt_(global/list)` to avoid any possible confusion with `--no-prompt` functionality.

File: runtime/permissions/mod.rs
Comment (iuioiua) [inline runtime/permissions/mod.rs:2647]: Should we now expect the right side of some of these assertions to be `PermissionState::GrantedPartial`? I expected it.

Also, if my understanding of `UnaryPermission.prompt` is correct, there shouldn't be any `PermissionState::Prompt` here. Is this right?

Comment (nayeemrmn) [reply runtime/permissions/mod.rs:2647]: > Should we now expect the right side of some of these assertions to be `PermissionState::GrantedPartial`? I expected it.

Nope, revoke behaviour is upwards recursing as it was before and is unaffected when there are no `--deny` flags. See https://deno.land/manual@v1.35.0/runtime/permission_apis#revoke-permissions. There are two reasons for this:
- There was no such thing as `GrantedPartial` until this PR. `read: /foo` always implied `read: /foo/bar`.
- The use case people brought up were to `--allow-read=./`, do some things with `./<files>`, revoke access to `./` while preserving access to `./public/` and then dynamic import dependencies.

Though, the other interpretation of `revoke()` can be argued for with similar use cases. i.e. `--allow-read=./`, do some things with `./private/`, revoke access downwards to `./private/` while preserving access to `./<files>`. Maybe that could be a different API like `Deno.permissions.deny()`.

File: cli/tests/testdata/run/deny_some_permission_args.out
Comment (nayeemrmn) [inline cli/tests/testdata/run/deny_some_permission_args.out:1]: I think this should be represented as `{ state: "granted", partial: true }` to not break compatibility with `{ state: "granted" }` and because for some use cases they are both acceptable results.

Comment (iuioiua) [reply cli/tests/testdata/run/deny_some_permission_args.out:1]: Ah! Yes. Fix incoming.

**Comment by iuioiua** [comment]
I'm unsure what's causing these errors. Specifically, running `cargo test integration::run::permissions::_061_permissions_request_sync` fails. However, manually running the test with `cargo run run --quiet cli/tests/testdata/run/061_permissions_request_sync.ts` and mimicking the input works fine. This is the case for every input (`y`, `n` and `A`) combination.

**Comment by nayeemrmn** [comment]
> running `cargo test integration::run::permissions::_061_permissions_request_sync` fails

Two things:
1. There's a bug on main that we unknowing fixed in this PR:
    ```ts
    Deno.permissions.request({ name: "read" });
    ```
    ```
    ┌ ⚠️  Deno requests read access.
    ├ Requested by `Deno.permissions.query()` API.
    ...
    ```
    Now the API name is corrected:
    ```
    ┌ ⚠️  Deno requests read access.
    ├ Requested by `Deno.permissions.request()` API.
    ...
    ```
    So we have to fix those expectations in the tests.
2. I forgot to preserve the `display_path` thing for request prompts (though I'm not sure why we need it there). Fix: https://github.com/nayeemrmn/deno/commit/e0316bcec3383946ae663cbfc3cd27f283972247.

**Comment by iuioiua** [comment]
Good catches! I've corrected the tests and applied your other suggestions. However, tests are still failing on my machine.

**Comment by iuioiua** [comment]
Hm... Seems that network permission queries are containing the network port unnecessarily even though `impl fmt::Display for NetDescriptor` is implemented correctly.

**Comment by nayeemrmn** [comment]
> Hm... Seems that network permission queries are containing the network port unnecessarily even though `impl fmt::Display for NetDescriptor` is implemented correctly.

https://github.com/nayeemrmn/deno/commit/3a7fb839b24e0d8925d17ac139c135e00672a806

**Comment by nayeemrmn** [comment]
https://github.com/nayeemrmn/deno/commit/b78010d290c662e6af9a364519e568cb51e02ffd I think this fixes the remaining tests. Besides fixes, it changes the net prompt text from `requests network access` to `requests net access`, since that was the only one not using the abbreviated flag name there. Other than that unary permission `description`s were unused so I removed them.

**Comment by iuioiua** [comment]
Thanks @nayeemrmn! This PR is now ready for review.

File: runtime/js/10_permissions.js
Comment (nayeemrmn) [inline runtime/js/10_permissions.js:147]: This breaks `onchange`. We need an early return at this selection level or the cached value will be displaced by the code below. I have a test case (which we should probably add):
```ts
const status = await Deno.permissions.query({ name: "read", path: "foo" });
console.log(status.state);
status.onchange = () => console.log(status.state);
await Deno.permissions.request({ name: "read", path: "foo" }); // y
await Deno.permissions.revoke({ name: "read", path: "foo" });

// Expected output:
//
// prompt
// ✅ Granted read access to "foo".
// granted
// prompt
```

File: runtime/ops/permissions.rs
Comment (nayeemrmn) [inline runtime/ops/permissions.rs:79]: We could factor this out to `impl From<PermissionState> for PermissionStatus {}`

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [reply runtime/permissions/mod.rs:235]: > This is a bit strange. Shouldn't it be `self > other`?

This is the default implementation which specifies no _stronger than_ relationship. But since it's a non-strict _stronger than_, the only time it's true is when the descriptors are the same.

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [inline runtime/permissions/mod.rs:258]: We should rework these names since 'denied' and 'no to prompt' aren't really different.

How about `denied` -> `flag_denied` and `no_to_prompt` -> `request_denied`?

**Review by nayeemrmn — COMMENTED** [review]
One correctness issue with `onchange` and some stylistic stuff

**Comment by nayeemrmn** [comment]
https://github.com/denoland/deno/issues/18804#issuecomment-1545167941 Also we have to change APIs like `Deno.watchFs()` , `Deno.remove(_, { recursive: true })` and `Deno.readDir()` to check non-partial permissions. Add `PermissionContainer::check_{read,write}_non_partial()` and replace the permission checks on those ops with these.

**We need to carefully check any APIs which do any sort of recursive access, add tests for when `--deny` flags are specified within their scopes and ensure the `--deny` flags aren't ignored.**

File: runtime/permissions/mod.rs
Comment (iuioiua) [reply runtime/permissions/mod.rs:258]: Good idea. How about `no_to_prompt` -> `prompt_denied`?

File: cli/tsc/dts/lib.deno.ns.d.ts
Comment (iuioiua) [reply cli/tsc/dts/lib.deno.ns.d.ts:4373]: Yeah, I was thinking that too. It's actually also not entirely correct, as `partial` is only ever `true`. Will remove.

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [reply runtime/permissions/mod.rs:258]: Yeah that's better 👍

**Comment by nayeemrmn** [comment]
@iuioiua List of remaining APIs that should use non partial permission checks:
```ts
Deno.readDirSync()
Deno.rename()
Deno.renameSync()
Deno.link()
Deno.linkSync()
```
The checks on `Deno.rename()` and `Deno.link()` are there to ensure that moving paths can't be used to bypass permission checks, all of these apply recursively.

You can search for `\.check_read|\.check_write` across the repo find the checks and double check the list.

**Comment by iuioiua** [comment]
> [#18804 (comment)](https://github.com/denoland/deno/issues/18804#issuecomment-1545167941) Also we have to change APIs like `Deno.watchFs()` , `Deno.remove(_, { recursive: true })` and `Deno.readDir()` to check non-partial permissions. Add `PermissionContainer::check_{read,write}_non_partial()` and replace the permission checks on those ops with these.

As per offline discussions with Nayeem, we've renamed `check_non_partial()` to the default `check()`, making it stricter and requiring fully granted permissions. The old `check()` has been renamed to `check_partial()` which passes when permissions are fully or partially granted. I've also taken the liberty, in the last commit, to make use of `perms.ffi.check()` more consistent with its path-related permission counterparts, `perms.read.check()` and `perms.write.check()`.

> We need to carefully check any APIs which do any sort of recursive access, add tests for when `--deny` flags are specified within their scopes and ensure the `--deny` flags aren't ignored.

`permissions::tests::test_query` takes decent care of this, no?

**Comment by nayeemrmn** [comment]
> > We need to carefully check any APIs which do any sort of recursive access, add tests for when `--deny` flags are specified within their scopes and ensure the `--deny` flags aren't ignored.
> 
> `permissions::tests::test_query` takes decent care of this, no?

Well, either way now that we've renamed the default to be the safer one I'm not as concerned about regressions there.

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [inline runtime/permissions/mod.rs:397]:
```suggestion
  fn is_flag_denied(&self, desc: &Option<T>) -> bool {
```

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [inline runtime/permissions/mod.rs:401]:
```suggestion
  fn is_prompt_denied(&self, desc: &Option<T>) -> bool {
```

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [inline runtime/permissions/mod.rs:411]:
```suggestion
  fn is_partial_flag_denied(&self, desc: &Option<T>) -> bool {
```

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [inline runtime/permissions/mod.rs:433]:
```suggestion
  fn insert_prompt_denied(&mut self, desc: Option<T>) {
```

**Review by nayeemrmn — COMMENTED** [review]
Some missed renames, looks good otherwise

**Review by nayeemrmn — APPROVED** [review]
LGTM!

**Comment by iuioiua** [comment]
IMO, there should also be a `--deny-all`. It'd be an alias for `--no-prompt`.

**Comment by nayeemrmn** [comment]
> IMO, there should also be a `--deny-all`. It'd be an alias for `--no-prompt`.

Let's put that off for now, I think the primary use case we unlock with this is carve-outs like `--allow-read=foo/ --deny-read=foo/secret`. Each global flag we add like `--deny-all` will add a little overhead and IMO there should be a good use for it.

**Comment by iuioiua** [comment]
My opinion isn't strong, but the logic for `--deny-all` pretty much already exists. We'd just have to link it up to the `--no-prompt` logic, no?

**Comment by nayeemrmn** [comment]
Doesn't seem so, `--deny` flags take precedent over and override `--allow` flags whereas `--no-prompt` only applies to ungranted permissions.

**Comment by iuioiua** [comment]
If that's the case, agreed.

File: cli/args/flags.rs
Comment (mmastrac) [inline cli/args/flags.rs:4823]: Given the default below, these can likely be omitted.

**Review by mmastrac — COMMENTED** [review]
Minor comments

**Comment by mmastrac** [comment]
@bartlomieju I am good with this PR but would be good to have your final approval.

File: runtime/permissions/mod.rs
Comment (bartlomieju) [inline runtime/permissions/mod.rs:52]: Is this enum's value somehow reachable to user code? Ie. will this be a breaking change from user code perspective that `GrantedPartial` has now value 1 instead of `Prompt`?

**Review by bartlomieju — COMMENTED** [review]
Some initial review, I'll be digging into this PR more over the coming days.

File: cli/tests/integration/run_tests.rs
Comment (nayeemrmn) [reply cli/tests/integration/run_tests.rs:582]: They already were, the prompt was wrong

File: runtime/permissions/mod.rs
Comment (nayeemrmn) [reply runtime/permissions/mod.rs:52]: I believe this is for serializing permissions between workers. It shouldn't leave process internals.

**Comment by iuioiua** [comment]
`stronger_than()` comment added.

**Comment by iuioiua** [comment]
`integration::run::stdio_streams_are_locked_in_permission_prompt` may be unrelated to this PR as it's also failing on the `main` branch on my local machine (macOS).

File: runtime/js/10_permissions.js
Comment (bartlomieju) [inline runtime/js/10_permissions.js:159]: @nayeemrmn is this correct? Shouldn't there only be a single `change` event emitted if both state and partial change? Or that can't really happen?

File: runtime/permissions/mod.rs
Comment (bartlomieju) [inline runtime/permissions/mod.rs:1]: Nice cleanup overall in this file 🚀

File: runtime/js/10_permissions.js
Comment (nayeemrmn) [reply runtime/js/10_permissions.js:159]: Good catch, 
```suggestion
    if (cachedObj.state !== rawStatus.state || cachedObj.partial !== rawStatus.partial) {
      cachedObj.state = rawStatus.state;
      cachedObj.partial = rawStatus.partial;
      cachedObj.status.dispatchEvent(
        new Event("change", { cancelable: false }),
      );
```

Comment (nayeemrmn) [reply runtime/js/10_permissions.js:159]: Though yes, partial status can't really change at all since you can't request `--deny`ed scopes, but I don't expect this code to be aware of that

**Review by bartlomieju — APPROVED** [review]
LGTM, fantastic work @iuioiua @nayeemrmn!

This feature will greatly expand capabilities of Deno's permissions system.

I will wait for some more reviews before merging this.

File: runtime/permissions/mod.rs
Comment (dsherret) [inline runtime/permissions/mod.rs:307]: It might be good to make this an enum with three options so it's more explanatory here and at the call sites.

**Review by dsherret — APPROVED** [review]
LGTM!

**Pull request merged**
